File: python/leetcode/dp/1223_dice_roll.py
"""
1223. Dice Roll Simulation (Medium)

A die simulator generates a random number from 1 to 6 for each roll. You introduced a constraint to the generator such that it cannot roll the number i more than rollMax[i] (1-indexed) consecutive times. 

Given an array of integers rollMax and an integer n, return the number of distinct sequences that can be obtained with exact n rolls.

Two sequences are considered different if at least one element differs from each other. Since the answer may be too large, return it modulo 10^9 + 7.

Example 1:

Input: n = 2, rollMax = [1,1,2,2,2,3]
Output: 34
Explanation: There will be 2 rolls of die, if there are no constraints on the die, there are 6 * 6 = 36 possible combinations. In this case, looking at rollMax array, the numbers 1 and 2 appear at most once consecutively, therefore sequences (1,1) and (2,2) cannot occur, so the final answer is 36-2 = 34.

Example 2:

Input: n = 2, rollMax = [1,1,1,1,1,1]
Output: 30
Example 3:

Input: n = 3, rollMax = [1,1,1,2,2,3]
Output: 181

Constraints:

1 <= n <= 5000
rollMax.length == 6
1 <= rollMax[i] <= 15
"""

class Solution(object):
    def dieSimulator(self, n, rollMax):
        """
        :type n: int
        :type rollMax: List[int]
        :rtype: int
        """
        memo = {}
        MOD = 10 ** 9 + 7
        for i in range(6):
            memo[i] = [0] * rollMax[i]
            memo[i][0] = 1

        for k in range(n-1):
            new_memo = {}

            tmp_memo = {}
            for key in memo:
                tmp_memo[key] = sum(memo[key])
            total = sum(tmp_memo.values())

            for key in memo:
                new_memo[key] = [0] + memo[key][:-1]
                new_memo[key][0] = (total - tmp_memo[key]) % MOD
            memo = new_memo
        
        result = 0
        for key in memo:
            result += sum(memo[key])
        return result % MOD


if __name__ == "__main__":
    a = Solution()
    print(a.dieSimulator(2, [1,1,2,2,2,3]))
    print(a.dieSimulator(2, [1,1,1,1,1,1]))
    print(a.dieSimulator(3, [1,1,1,2,2,3]))

File: cc/leetcode/dp/1223_dice_roll.cc
class Solution {
public:
    int dieSimulator(int n, vector<int>& rollMax) {
        vector<vector<int>> memo;
        int MOD = pow(10, 9) + 7;
        for (int i = 0; i < 6; ++i) {
            vector<int> tmp(rollMax[i], 0);
            memo.push_back(tmp);
            memo[i][0] = 1;
        }
        
        for (int k = 0; k < n-1; ++k) {
            vector<vector<int>> new_memo;
            
            vector<int> memo_sum(6, 0);
            int total = 0;
            for (int i = 0; i < 6; ++i) {
                for (int j = 0; j < rollMax[i]; ++j)
                    memo_sum[i] = (memo_sum[i] + memo[i][j]) % MOD;
                total = (total + memo_sum[i]) % MOD;
            }
            
            for (int i = 0; i < 6; ++i) {
                int c = rollMax[i];
                vector<int> tmp(c, 0);
                new_memo.push_back(tmp);
                for (int j = 0; j < c-1; ++j)
                    new_memo[i][j+1] = memo[i][j];
                new_memo[i][0] = (total - memo_sum[i]) % MOD;
            }
            memo = new_memo;
        }
        
        int result = 0;
        for (int i = 0; i < 6; ++i) {
            for (int j = 0; j < memo[i].size(); ++j)
                result = (result + memo[i][j]) % MOD;
        }
        if (result >= 0)
            return result;
        else
            return MOD + result;
    }
};
